packages/smart-vaults/src/vault/SmartVault.sol

```diff
@@ -394,4 +435,34 @@ contract SmartVault is IAccount, Ownable, UUPSUpgradeable, MultiSignerAuth, ERC1
     function _getSignatureType(bytes1 signatureType_) internal pure returns (SignatureTypes) {
         return SignatureTypes(uint8(signatureType_));
     }
+
+    function _verifyGasLimits(
+        LightUserOpGasLimits memory gasLimits_,
+        PackedUserOperation calldata userOp_
+    )
+        internal
+        pure
+    {
+        (uint256 userOpMaxPriorityFeePerGas,) = UserOperationLib.unpackUints(userOp_.gasFees);
+        (uint256 verificationGasLimit, uint256 callGasLimit) = UserOperationLib.unpackUints(userOp_.accountGasLimits);
+
+        if (
+            userOpMaxPriorityFeePerGas > gasLimits_.maxPriorityFeePerGas || callGasLimit > gasLimits_.callGasLimit
+                || userOp_.preVerificationGas > gasLimits_.preVerificationGas
+                || verificationGasLimit > gasLimits_.verificationGasLimit
+        ) revert InvalidGasLimits();
+
+        if (userOp_.paymasterAndData.length > 0) {
+            (address paymaster, uint256 paymasterVerificationGasLimit, uint256 paymasterPostOpGasLimit) =
+                UserOperationLib.unpackPaymasterStaticFields(userOp_.paymasterAndData);
+
+            if (
+                gasLimits_.paymaster != paymaster
```

Comment:
is there any reason we might want the final signer to have some flexibility over picking a paymaster? i can't say that really makes sense to me but just throwing it out there..

Reply:
I have one potential argument in favor of keeping the `gasLimits_.paymaster` <> `paymaster` check:

When the light signers set their `paymasterVerificationGasLimit` and `paymasterPostOpGasLimit` values, they would have a specific paymaster in mind, and they'd set the two values based on how much gas that paymaster uses. If the final signer changes to a different paymaster that uses less gas, the light signers probably would have preferred setting a lower upper-bound, so in their eyes they've accidentally given more leeway for wasting gas